File: nodejs/src/bot/handlers/updateHandlers.ts
/**
 * /version  — 查看當前程式版本
 * /update   — 檢查更新 + 確認後執行 Blue-Green 更新 + 重啟
 * /rollback — 回滾到上一個備份版本 + 重啟
 * /restart  — 立即重啟進程
 *
 * 以上指令僅限管理員使用。
 */

import type { Bot, Context } from "grammy";
import { InlineKeyboard } from "grammy";
import type { ConversationFlavor } from "@grammyjs/conversations";
import { isAdmin } from "../filters.js";
import {
  getCurrentVersion,
  fetchAndCheckUpdate,
  performUpdate,
  restartProcess,
  isWorkingDirClean,
  getBackupList,
  rollbackAndRestart,
} from "../../updater.js";

type MyContext = Context & ConversationFlavor;

// ========================
// 輔助函數
// ========================

/** 格式化日期顯示 */
function formatDate(isoDate: string): string {
  if (!isoDate) return "—";
  try {
    const d = new Date(isoDate);
    return d.toLocaleString("zh-TW", { timeZone: "Asia/Taipei" });
  } catch {
    return isoDate;
  }
}

// ========================
// /version — 查看當前版本
// ========================

async function versionCommand(ctx: MyContext): Promise<void> {
  try {
    const version = getCurrentVersion();
    await ctx.reply(
      `📦 *目前版本*\n\n` +
      (version.tag ? `🏷️ Release：\`${version.tag}\`\n` : "") +
      `🔖 Commit：\`${version.hash}\`\n` +
      `📝 訊息：${version.message}\n` +
      `🕐 時間：${formatDate(version.date)}`,
      { parse_mode: "Markdown" },
    );
  } catch (err: any) {
    await ctx.reply(`❌ 取得版本失敗：${err.message}`);
  }
}

// ========================
// /update — 檢查更新 + 確認執行
// ========================

async function updateCommand(ctx: MyContext): Promise<void> {
  try {
    await ctx.reply("⏳ 正在檢查更新...");

    const result = await fetchAndCheckUpdate();

    if (!result.hasUpdate) {
      await ctx.reply(
        `✅ 已是最新版本！\n\n` +
        (result.current.tag ? `🏷️ Release：\`${result.current.tag}\`\n` : "") +
        `🔖 Commit：\`${result.current.hash}\`\n` +
        `📝 ${result.current.message}\n` +
        `🕐 ${formatDate(result.current.date)}`,
        { parse_mode: "Markdown" },
      );
      return;
    }

    // 有更新可用
    let msg =
      `🔄 *有新版本可用！*\n\n` +
      `📍 *當前版本*\n` +
      (result.current.tag ? `🏷️ \`${result.current.tag}\`\n` : "") +
      `🔖 \`${result.current.hash}\`\n` +
      `📝 ${result.current.message}\n` +
      `🕐 ${formatDate(result.current.date)}\n\n`;

    // 顯示 GitHub Release 資訊
    if (result.latestRelease) {
      msg +=
        `🆕 *GitHub 最新 Release*\n` +
        `🏷️ \`${result.latestRelease.tag}\`` +
        (result.latestRelease.prerelease ? " *(預發布)*" : " *(穩定版)*") + `\n` +
        `📝 ${result.latestRelease.name}\n` +
        `🕐 ${formatDate(result.latestRelease.publishedAt)}\n`;

      if (result.current.tag && result.latestRelease.tag) {
        msg += `🔗 [查看 Release](${result.latestRelease.htmlUrl})\n`;
      }
      msg += "\n";
    }

    // 顯示落後 commit 數量
    if (result.commitsBehind > 0) {
      msg += `📊 落後 ${result.commitsBehind} 個提交\n`;
      const display = result.newCommits.slice(0, 10);
      if (display.length > 0) {
        msg += `\n📜 *新增提交：*\n${display.join("\n")}`;
        if (result.newCommits.length > 10) {
          msg += `\n... 還有 ${result.newCommits.length - 10} 條`;
        }
      }
    }

    // 檢查工作目錄
    const clean = isWorkingDirClean();
    if (!clean) {
      msg += "\n\n⚠️ *警告：工作目錄有未提交的更改！*\n更新將改用 tarball 下載方式。";
    }

    const keyboard = new InlineKeyboard()
      .text("✅ 確認更新", "update_confirm")
      .text("❌ 取消", "update_cancel");

    await ctx.reply(msg, {
      parse_mode: "Markdown",
      reply_markup: keyboard,
    });
  } catch (err: any) {
    await ctx.reply(`❌ 檢查更新失敗：${err.message}`);
  }
}

// ========================
// /rollback — 回滾到上一個備份版本
// ========================

async function rollbackCommand(ctx: MyContext): Promise<void> {
  try {
    const backups = getBackupList();

    if (backups.length === 0) {
      await ctx.reply("📦 沒有可用的備份版本。\n\n需要先執行過 /update 才能回滾。");
      return;
    }

    const lines = backups.slice(0, 5).map((b, i) => {
      const time = b.createdAt.toLocaleString("zh-TW", { timeZone: "Asia/Taipei" });
      return `${i === 0 ? "🟢" : "⚪"} \`${b.name}\`\n   🕐 ${time}`;
    });

    const keyboard = new InlineKeyboard()
      .text("✅ 確認回滾", "rollback_confirm")
      .text("❌ 取消", "rollback_cancel");

    await ctx.reply(
      `📦 *可回滾的備份版本*\n\n${lines.join("\n\n")}\n\n` +
      `⚠️ 將回滾到最新的備份（🟢）。當前版本會自動備份。`,
      { parse_mode: "Markdown", reply_markup: keyboard },
    );
  } catch (err: any) {
    await ctx.reply(`❌ 取得備份列表失敗：${err.message}`);
  }
}

async function handleRollbackConfirm(ctx: MyContext): Promise<void> {
  try {
    await ctx.answerCallbackQuery();
    await ctx.editMessageText("⏳ 正在回滾...");
    rollbackAndRestart();
  } catch (err: any) {
    try {
      await ctx.answerCallbackQuery();
    } catch { /* ignore */ }
    await ctx.editMessageText(`❌ 回滾失敗：${err.message}`);
  }
}

async function handleRollbackCancel(ctx: MyContext): Promise<void> {
  try {
    await ctx.answerCallbackQuery({ text: "已取消回滾" });
    await ctx.editMessageText("🚫 已取消回滾。");
  } catch { /* ignore */ }
}

// ========================
// /restart — 立即重啟
// ========================

async function restartCommand(ctx: MyContext): Promise<void> {
  await ctx.reply(
    "🔄 *正在重啟進程...*\n\nBot 將在 2 秒後重新上線。",
    { parse_mode: "Markdown" },
  );

  console.log("[restart] 由管理員觸發重啟");
  restartProcess(2000);
}

// ========================
// Callback Query 處理
// ========================

async function handleUpdateConfirm(ctx: MyContext): Promise<void> {
  try {
    await ctx.answerCallbackQuery();
    await ctx.editMessageText(
      "⏳ 正在更新程式碼...\n\nBlue-Green 更新中：下載 → npm install → 編譯 → 原子交換。",
    );

    const result = await performUpdate();

    if (result.success) {
      const methodText =
        result.method === "blue-green" ? "🔄 Blue-Green 原子交換"
        : result.method === "tarball" ? "📦 tarball 下載"
        : "📥 git pull";
      await ctx.editMessageText(
        `✅ ${result.message}\n\n` +
        `🔧 更新方式：${methodText}\n\n` +
        `🔄 正在重啟進程...\nBot 將在 2 秒後重新上線。`,
      );

      console.log("[update] 更新成功，正在重啟...");
      restartProcess(2000);
    } else {
      await ctx.editMessageText(`❌ ${result.message}`);
    }
  } catch (err: any) {
    try {
      await ctx.answerCallbackQuery();
    } catch { /* ignore */ }
    await ctx.editMessageText(`❌ 更新失敗：${err.message}`);
  }
}

async function handleUpdateCancel(ctx: MyContext): Promise<void> {
  try {
    await ctx.answerCallbackQuery({ text: "已取消更新" });
    await ctx.editMessageText("🚫 已取消更新。");
  } catch { /* ignore */ }
}

// ========================
// 註冊 Handler
// ========================

export function registerUpdateHandlers(bot: Bot<MyContext>): void {
  // Admin-only wrapper
  const adminOnly = (
    handler: (ctx: MyContext) => Promise<void>,
  ) => async (ctx: MyContext) => {
    if (!isAdmin(ctx)) {
      await ctx.reply("⛔ 此指令僅限管理員使用。");
      return;
    }
    await handler(ctx);
  };

  bot.command("version", adminOnly(versionCommand));
  bot.command("update", adminOnly(updateCommand));
  bot.command("rollback", adminOnly(rollbackCommand));
  bot.command("restart", adminOnly(restartCommand));

  // Callback queries
  bot.callbackQuery("update_confirm", adminOnly(handleUpdateConfirm));
  bot.callbackQuery("update_cancel", adminOnly(handleUpdateCancel));
  bot.callbackQuery("rollback_confirm", adminOnly(handleRollbackConfirm));
  bot.callbackQuery("rollback_cancel", adminOnly(handleRollbackCancel));
}

File: python/bot/handlers/update_handlers.py
"""
Bot handlers for in-built update commands.

/version  — Show current program version (admin only)
/update   — Check for updates + confirm via inline keyboard (admin only)
/restart  — Restart the process immediately (admin only)
"""

from __future__ import annotations

import logging

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.ext import Application, CallbackQueryHandler, CommandHandler, ContextTypes, filters

from bot.filters import admin_filter
from config import Config
from updater import (
    fetch_and_check_update,
    get_current_version,
    is_working_dir_clean,
    perform_update,
    restart_process,
)

logger = logging.getLogger(__name__)


# ============================================================
# 輔助函數
# ============================================================


def _format_date(iso_date: str) -> str:
    """格式化日期顯示"""
    if not iso_date:
        return "—"
    try:
        from datetime import datetime
        dt = datetime.fromisoformat(iso_date)
        return dt.strftime("%Y-%m-%d %H:%M:%S")
    except Exception:
        return iso_date


# ============================================================
# /version — 查看當前版本
# ============================================================


async def version_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Show current program version."""
    try:
        version = get_current_version()
        tag_line = f"🏷️ Release：`{version.tag}`\n" if version.tag else ""
        await update.message.reply_text(
            f"📦 *目前版本*\n\n"
            f"{tag_line}"
            f"🔖 Commit：`{version.hash}`\n"
            f"📝 訊息：{version.message}\n"
            f"🕐 時間：{_format_date(version.date)}",
            parse_mode="Markdown",
        )
    except Exception as e:
        await update.message.reply_text(f"❌ 取得版本失敗：{e}")


# ============================================================
# /update — 檢查更新 + 確認執行
# ============================================================


async def update_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Check for updates and show confirmation inline keyboard."""
    try:
        await update.message.reply_text("⏳ 正在檢查更新...")

        result = fetch_and_check_update()

        if not result.has_update:
            tag_line = f"🏷️ Release：`{result.current.tag}`\n" if result.current.tag else ""
            await update.message.reply_text(
                f"✅ 已是最新版本！\n\n"
                f"{tag_line}"
                f"🔖 Commit：`{result.current.hash}`\n"
                f"📝 {result.current.message}\n"
                f"🕐 {_format_date(result.current.date)}",
                parse_mode="Markdown",
            )
            return

        # 有更新可用
        tag_line = f"🏷️ `{result.current.tag}`\n" if result.current.tag else ""
        msg = (
            f"🔄 *有新版本可用！*\n\n"
            f"📍 *當前版本*\n"
            f"{tag_line}"
            f"🔖 `{result.current.hash}`\n"
            f"📝 {result.current.message}\n"
            f"🕐 {_format_date(result.current.date)}\n\n"
        )

        # 顯示 GitHub Release 資訊
        if result.latest_release:
            rel = result.latest_release
            pre_label = " *(預發布)*" if rel.prerelease else " *(穩定版)*"
            msg += (
                f"🆕 *GitHub 最新 Release*\n"
                f"🏷️ `{rel.tag}`{pre_label}\n"
                f"📝 {rel.name}\n"
                f"🕐 {_format_date(rel.published_at)}\n"
            )
            if result.current.tag and rel.tag:
                msg += f"🔗 [查看 Release]({rel.html_url})\n"
            msg += "\n"

        # 顯示落後 commit 數量
        if result.commits_behind > 0:
            msg += f"📊 落後 {result.commits_behind} 個提交\n"
            display = result.new_commits[:10]
            if display:
                msg += f"\n📜 *新增提交：*\n{chr(10).join(display)}"
                if len(result.new_commits) > 10:
                    msg += f"\n... 還有 {len(result.new_commits) - 10} 條"

        # 檢查工作目錄
        if not is_working_dir_clean():
            msg += "\n\n⚠️ *警告：工作目錄有未提交的更改！*\n更新將改用 tarball 下載方式。"

        keyboard = InlineKeyboardMarkup([
            [
                InlineKeyboardButton("✅ 確認更新", callback_data="update_confirm"),
                InlineKeyboardButton("❌ 取消", callback_data="update_cancel"),
            ]
        ])

        await update.message.reply_text(
            msg,
            parse_mode="Markdown",
            reply_markup=keyboard,
        )
    except Exception as e:
        await update.message.reply_text(f"❌ 檢查更新失敗：{e}")


# ============================================================
# /restart — 立即重啟
# ============================================================


async def restart_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Restart the process immediately."""
    await update.message.reply_text(
        "🔄 *正在重啟進程...*\n\nBot 將在 2 秒後重新上線。",
        parse_mode="Markdown",
    )
    logger.info("[restart] 由管理員觸發重啟")
    await restart_process(2.0)


# ============================================================
# Callback Query 處理
# ============================================================


async def handle_update_confirm(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Handle the update confirmation callback."""
    query = update.callback_query
    # Admin check for callback queries (no filters= param for CallbackQueryHandler)
    if not update.effective_user or update.effective_user.id != Config.ADMIN_ID:
        await query.answer(text="⛔ 此操作僅限管理員。", show_alert=True)
        return
    try:
        await query.answer()
        await query.edit_message_text(
            "⏳ 正在更新程式碼...\n\n嘗試 git pull，失敗則改用 tarball 下載。"
        )

        result = perform_update()

        if result.success:
            method_text = "📦 tarball 下載" if result.method == "tarball" else "📥 git pull"
            await query.edit_message_text(
                f"✅ {result.message}\n\n"
                f"🔧 更新方式：{method_text}\n\n"
                f"🔄 正在重啟進程...\nBot 將在 2 秒後重新上線。"
            )
            logger.info("[update] 更新成功，正在重啟...")
            await restart_process(2.0)
        else:
            await query.edit_message_text(f"❌ {result.message}")
    except Exception as e:
        try:
            await query.answer()
        except Exception as e2:
            logger.debug("query.answer failed in update error handler: %s", e2)
        try:
            await query.edit_message_text(f"❌ 更新失敗：{e}")
        except Exception as e2:
            logger.warning("Failed to notify user about update error: %s", e2)


async def handle_update_cancel(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Handle the update cancel callback."""
    query = update.callback_query
    if not update.effective_user or update.effective_user.id != Config.ADMIN_ID:
        await query.answer(text="⛔ 此操作僅限管理員。", show_alert=True)
        return
    try:
        await query.answer(text="已取消更新")
        await query.edit_message_text("🚫 已取消更新。")
    except Exception as e:
        logger.debug("Failed to acknowledge update cancel: %s", e)


# ============================================================
# 註冊 Handler
# ============================================================


def register_update_handlers(app: Application) -> None:
    """Register all update-related command handlers (admin only)."""
    app.add_handler(CommandHandler("version", version_command, filters=admin_filter))
    app.add_handler(CommandHandler("update", update_command, filters=admin_filter))
    app.add_handler(CommandHandler("restart", restart_command, filters=admin_filter))

    # Callback queries
    app.add_handler(CallbackQueryHandler(handle_update_confirm, pattern="^update_confirm$"))
    app.add_handler(CallbackQueryHandler(handle_update_cancel, pattern="^update_cancel$"))
